rusk/src/lib/services/version.rs

```
@@ -0,0 +1,185 @@
+    >;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: hyper::Request<Body>) -> Self::Future {
+        // This is necessary because tonic internally uses
+        // `tower::buffer::Buffer`. See https://github.com/tower-rs/tower/issues/547#issuecomment-767629149
+        // for details on why this is necessary
+        let clone = self.inner.clone();
+        let mut inner = std::mem::replace(&mut self.inner, clone);
+
+        Box::pin(async move {
+            let mut response = inner.call(req).await?;
+            let headers = response.headers_mut();
+            headers.append("x-rusk-version", HeaderValue::from_static(VERSION));
+            Ok(response)
+        })
+    }
+}
+
+/// Checks incoming requests for compatibility with running
+/// Rusk version using an Interceptor.
+#[derive(Clone)]
+pub struct CompatibilityInterceptor;
+
+impl Interceptor for CompatibilityInterceptor {
+    fn call(
+        &mut self,
+        request: tonic::Request<()>,
+    ) -> Result<tonic::Request<()>, Status> {
+        // attempt to extract `x-rusk-version` header metadata
+        let metadata = request.metadata();
+        match metadata.get("x-rusk-version") {
+            Some(header_v) => {
+                if let Ok(req_v) = header_v.to_str() {
+                    if let Ok(res) = is_compatible(req_v) {
+                        match res {
+                            true => (),
+                            false => {
+                                return Err(Status::unavailable(
+                                    "version not supported",
+                                ))
+                            }
+                        }
+                    }
+                }
+            }
+            None => {
+                return Err(Status::unavailable("incompatible client version"))
```

I think the error is misleading, it should be something like `missing "x-rusk-version" header`.
Furthermore, should we raise an error on missing header only if a strict-check flag is enabled?
This way we can smoothly update existing callers

---

Or maybe simply skip the check if no header is passed

---

We decided not to skip to enforce clients to update themselves... cc @ZER0